### PR TITLE
Add entrypoint for plotting backends

### LINKED ISSUE #3553: ENH: Plotting backend options
Since pandas has implemented entry_points based plotting backends, it seems reasonable that xarray would do the same. This would make it even easier to produce holoviews plots (rendered in bokeh via hvplot), by using the `plot` method rather than by importing hvplot directly.

#### Example

```python
import xarray as xr
air = xr.tutorial.open_dataset('air_temperature').load().air
xr.options.plotting.backend = 'holoviews'

air.isel(time=500).plot()
```
Screen Shot 2019-11-20 at 12 44 26 PM

### PR DESCRIPTION
**Pull request by andersy005**
Libraries, including `xarray`, register backends via `entrypoints`. Xarray adds an EntryPoint "group" in its `setup.py`. Third party libraries add items to that group in their `setup.pys`. 

Xarray and the third-party library don't need to talk to each other directly to register a backend. They talk through `pkg_resources`.

For instance, `hvplot/holoviews` would add an EntryPoint for `xarray_plotting_backends` in its setup.py:

```python
entry_points={"xarray_plotting_backends": ['holoviews = hvplot:plotting']}
```

and xarray users should be able to achieve what @jsignell proposed in #3553:

> ```python
> import xarray as xr
> air = xr.tutorial.open_dataset('air_temperature').load().air
> xr.set_options(plotting_backend='holoviews')
> 
> air.isel(time=500).plot()
> ```
> 
> Screen Shot 2019-11-20 at 12 44 26 PM

 - [ ] Closes #3553
 - [x] Tests added
 - [x] Passes `black . && mypy . && flake8`
 - [ ] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

### CODE REVIEW

**Comment by dcherian** [comment]
EXCITING!

**Comment by crusaderky** [comment]
if you merge from master you should get all green on the upstream-dev test now

**Comment by dcherian** [comment]
@andersy005 can you summarize what's needed on the holoviews and xarray side for this to work?

**Comment by andersy005** [comment]
> can you summarize what's needed on the holoviews and xarray side for this to work?

A while back I opened this issue https://github.com/holoviz/hvplot/issues/392. I am about to make modifications to hvplot/holoviews locally to test that this PR is going to work as expected. Once I have a better understanding of what's needed, I will report back.

**Comment by andersy005** [comment]
After looking at Pandas source code, it's my understanding that 
- Pandas provides a well defined [plotting API](https://github.com/pandas-dev/pandas/blob/v1.0.0/pandas/plotting/_core.py) 
- At runtime, Pandas figures out the right backend to use via [`_get_plot_backend()`](https://github.com/pandas-dev/pandas/blob/v1.0.0/pandas/plotting/_core.py#L1645) and then dispatches the plotting functionality to this backend.

This approach makes interfacing with Pandas objects easy as long as the plotting library adheres to this plotting API. Currently, Xarray's plotting API is closely tied to matplotlib. To support entrypoint for other plotting backend will likely require defining something similar to [Pandas's plotting API](https://github.com/pandas-dev/pandas/blob/v1.0.0/pandas/plotting/_core.py). **Is my interpretation correct?**

Pinging @TomAugspurger, @jsignell and @jbednar as they may have ideas on how to proceed with this implementation.

**Comment by TomAugspurger** [comment]
> Is my interpretation correct?
 
Yep, that's the basic idea. Every call to `DataFrame.plot.<kind>` begins with a [check for the active backend](https://github.com/pandas-dev/pandas/blob/v1.0.0/pandas/plotting/_core.py#L767). Based on the configured value, we the correct backend, make the call, and return the result.

**Comment by philippjfr** [comment]
@andersy005 Sorry I never followed up here. Trying to get out a new hvPlot release very soon. What's the current status here?  @jsignell do you have any comments?

**Comment by philippjfr** [comment]
To be clear, I'm just checking whether there's anything that I can or need to do in hvPlot so this can proceed smoothly.

**Comment by andersy005** [comment]
> What's the current status here?

I put this PR on a back burner :). As I pointed out in https://github.com/pydata/xarray/pull/3640#issuecomment-580413432, it is my understand that xarray will need to define a minimal plotting API for this entrypoint feature to work properly. Would this plotting API be something xarray devs would be willing to have/support? Cc @shoyer @dcherian @jhamman 


I could be missing something here. So please let me know if there is another way of getting the entrypoint feature to work without needing to defining a plotting API.

**Comment by andersy005** [comment]
Closing this because it has gone stale :(

**Comment by TomNicholas** [comment]
> it is my understand that xarray will need to define a minimal plotting API for this entrypoint feature to work properly. Would this plotting API be something xarray devs would be willing to have/support?

Do we not want to discuss this in the bi-weekly dev meetings perhaps? (Apologies if it has been and I just missed it.) This seems like a good idea!

**Comment by andersy005** [comment]
> Do we not want to discuss this in the bi-weekly dev meetings perhaps?

👍🏽 I'm in favor of discussing this in the dev meetings... Being able to easily switch plotting backends would be a nice feature to have from a user's perspective.

**Comment by jsignell** [comment]
I'm sorry I missed pings on this. I am happy to join the dev meeting if that'd be helpful or I can just start paying better attention :roll_eyes:

**Comment by TomNicholas** [comment]
There is a [dev meeting this Wednesday](https://github.com/pydata/xarray/issues/4001) (occurring every fortnight), would be great to see you there!
